agents/agents-features/agents-features-tokenizer/src/commonMain/kotlin/ai/koog/agents/features/tokenizer/feature/MessageTokenizer.kt

```diff
@@ -0,0 +1,221 @@
+     * This feature integrates a message tokenizer into the agent pipeline, allowing for tokenization
+     * of input messages. It supports both caching and non-caching tokenization strategies based on the configuration.
+     */
+    public companion object Feature : AIAgentFeature<MessageTokenizerConfig, MessageTokenizer> {
+
+        /**
+         * A unique storage key used to identify the `MessageTokenizer` feature within the agent's feature storage.
+         * This key ensures that the `MessageTokenizer` instance can be retrieved or referenced
+         * when required during the lifecycle or operation of the agent.
+         */
+        override val key: AIAgentStorageKey<MessageTokenizer> =
+            AIAgentStorageKey("agents-features-tracing")
+
+        /**
+         * Creates and returns the initial configuration for the `MessageTokenizer` feature.
+         *
+         * @return A new instance of `MessageTokenizerConfig` containing the default configuration.
+         */
+        override fun createInitialConfig(): MessageTokenizerConfig = MessageTokenizerConfig()
+
+        /**
+         * Installs the MessageTokenizer feature into the given AI Agent pipeline.
+         *
+         * Configures and initializes the appropriate tokenizer (caching or on-demand)
+         * based on the provided configuration, and registers the MessageTokenizer
+         * feature into the pipeline.
+         *
+         * @param config The configuration used to customize the MessageTokenizer feature, including tokenizer settings and caching options.
+         * @param pipeline The AI Agent pipeline where the MessageTokenizer feature will be installed.
+         */
+        override fun install(
+            config: MessageTokenizerConfig,
+            pipeline: AIAgentPipeline,
+        ) {
+            val promptTokenizer = if (config.enableCaching)
+                CachingTokenizer(config.tokenizer)
+            else
+                OnDemandTokenizer(config.tokenizer)
+
+            val feature = MessageTokenizer(promptTokenizer)
+
+            pipeline.interceptContextAgentFeature(this) { feature }
+        }
+    }
+}
+
+/**
+ * Provides access to the `PromptTokenizer` instance used within the AI agent's context.
+ *
+ * This property retrieves the tokenizer from the agent's storage using the `MessageTokenizer.Feature`,
+ * which must be initialized in the pipeline's features. The `PromptTokenizer` allows
+ * for tokenization operations on prompts and messages during the agent's execution.
+ *
+ * It facilitates operations such as calculating token counts for messages and prompts,
+ * which are critical in managing and optimizing interactions with language models.
+ *
+ * Throws an exception if the `MessageTokenizer.Feature` is not available in the context.
+ */
+public val AIAgentContextBase.tokenizer: PromptTokenizer get() = feature(MessageTokenizer.Feature)!!.promptTokenizer
```

Comment:
This would throw NPE, which is not really descriptive. It's better to throw exception manually with more descriptive message